spec.md

```diff
@@ -1090,6 +1119,287 @@ message ControllerServiceCapability {
+
+This operation MUST be idempotent.
+If a snapshot corresponding to the specified snapshot `name` already exists and is compatible with the specified `volume_id` in the `CreateSnapshotRequest`, the Plugin MUST reply `0 OK` with the corresponding `CreateSnapshotResponse`.
+
+```protobuf
+message CreateSnapshotRequest {
+  // The ID of the volume to be snapshotted.
+  // This field is REQUIRED.
+  string volume_id = 1;
+
+  // The suggested name for the snapshot. This field is REQUIRED.
+  // It serves two purposes:
+  // 1) Idempotency - This name is generated by the CO to achieve
+  //    idempotency. If `CreateSnapshot` fails, the snapshot may or
+  //    may not be created. In this case, the CO may call
+  //    `CreateSnapshot` again, with the same name, to ensure the
+  //    snapshot exists. The Plugin should ensure that multiple
+  //    `CreateSnapshot` calls for the same name do not result in more
+  //    than one snapshot created corresponding to that name. If a
+  //    Plugin is unable to enforce idempotency, the CO's error recovery
+  //    logic could result in multiple (unused) snapshots being created.
+  // 2) Suggested name - Some storage systems allow callers to specify
+  //    an identifier by which to refer to the newly created snapshot.
+  //    If a storage system supports this, it can optionally use this
+  string name = 2;
+
+  // The description of the snapshot.
+  // This field is OPTIONAL.
+  string description = 3;
+
+  // Credentials used by Controller plugin to authenticate/authorize
+  // snapshot creation request.
+  // This field contains credential data, for example username and
+  // password. Each key must consist of alphanumeric characters, '-',
+  // '_' or '.'. Each value MUST contain a valid string. An SP MAY
+  // choose to accept binary (non-string) data by using a binary-to-text
+  // encoding scheme, like base64. An SP SHALL advertise the
+  // requirements for credentials in documentation. COs SHALL permit
+  // users to pass through the required credentials. This information is
+  // sensitive and MUST be treated as such (not logged, etc.) by the CO.
+  // This field is OPTIONAL.
+  map<string, string> controller_create_credentials = 4;
+
+  // NOTE: In the future, we may want to add “tags” to identify a
+  // snapshot. This is generated by CO internally and passed to the
+  // cloud storage backend through the plugin. This can serve the
+  // following purpose: Allow cloud storage admin to identify snapshots
+  // created by a specific CO.
+  // map<string, string> tags = 5;
+}
+
+message CreateSnapshotResponse {
+  // Contains all attributes of the newly created snapshot that are
+  // relevant to the CO along with information required by the Plugin
+  // to uniquely identify the snapshot. This field is REQUIRED.
+  Snapshot snapshot = 1;
+}
+
+// The information about a provisioned snapshot.
+message Snapshot {
```

Comment:
I like the idea of `Snapshot` containing ONLY those fields that won't change over time.

Reply:
Does that mean `status` shouldn't be in `Snapshot`? (re: `Snapshot` containing ONLY those fields that won't change over time).

Reply:
If we were to separate immutable fields from mutable fields, maybe  we should rename `GetSnapshot` to `GetSnapshotStatus`?

Reply:
How about we remove Status from the Snapshot response from CreateSnapshot and GetSnapshot and keep the Status only in the response from ListSnapshots?

1. CO calls CreateSnapshot. CreateSnapshot will block until snapshot is cut.
2. CO checks if REQUIRES_GET_SNAPSHOT is a capability supported by the plugin.  If yes, CO will call GetSnapshot.
3a. CO calls GetSnapshot. GetSnapshot will block until snapshot is uploaded.  When GetSnapshot returns, CO is done with snapshot creation and uploading.
3b. If REQUIRES_GET_SNAPSHOT is not supported, CO is done with snapshot creation without calling GetSnapshot.

CO can always call ListSnapshots with snapshot_id as the filter to find out the Status of a particular snapshot.   ListSnapshots does not block and will return with the current snapshot info.

Reply:
I think the different behaviors of `GetSnapshot` and `ListSnapshots` are quite unexpected. Normally GetSnapshot is the singular form of ListSnapshots. I don't expect it to function completely different (blocking vs non-blocking, reporting the error etc).

How about we rename `GetSnapshot` to e.g. `WaitForSnapshotComplete`?

Reply:
Agree that the naming is confusing.  There's still debate on whether GetSnapshot is needed.  Once that decision is made, we can decide whether renaming is needed.